### PR TITLE
Transpose adjacency matrices so source nodes are rows

### LINKED ISSUE #27: Are our adjacency matrices transposed?
Currently, our adjacency matrices for metaedges encode source nodes as columns and target nodes as rows. I'm starting to think this is backwards. Two observations:

1. to access a specific value, we must go `matrix[target_pos, source_pos]`. It's more natural to declare source then target.

2. In the proposed [DWWC implemenation](https://github.com/dhimmel/hetmech/blob/2ed28bb589caa909a85dcf979207d4dae827eb7d/hetmech/diffusion.py#L160), we have to do `dwwc_matrix = adj_mat @ dwwc_matrix` rather than `dwwc_matrix = dwwc_matrix @ adj_mat`. The later is preferable in that we could use `functools.reduce`.

@kkloste what do you think?

### PR DESCRIPTION
**Pull request by dhimmel**
Closes #27.

### CODE REVIEW

**Comment by dhimmel** [comment]
@kkloste the transposition I did in `hetmech/diffusion.py` of https://github.com/greenelab/hetmech/pull/32/commits/a5bd2bd62ce1be7f97b0e6be45c956552f31f948 gives different results in the diffusion notebook.

I'm assuming this is because I need to do a bit more than just switch the order of multiplication.

File: hetmech/diffusion.py [hunk 1]
```diff
@@ -90,7 +90,7 @@ def diffusion(
         adjacency_matrix = diffusion_step(
             adjacency_matrix, row_damping, column_damping)
 
-        node_scores = adjacency_matrix @ node_scores
+        node_scores = node_scores @ adjacency_matrix
```
Comment (dhimmel) [inline hetmech/diffusion.py]: @kkloste do we need node scores to be transposed as well?

**Comment by kkloste** [comment]
@dhimmel  Yes, we have to transpose the normalizations, too.

Suppose a metapath is a sequence of 4 meta edges, and we compute the diffusion by multiplying together the normalized diffusion matrices for each edge type:
P4 * P3 * P2 * P1

where the source nodes correspond to the columns of matrix P1, the target nodes correspond to the rows of matrix P4. In this context, what we've been doing to normalize is scale each row by that node's degree, then normalizing each column to sum to 1 (i.e. be "juice-preserving").

The desired new way of performing this computation is to have rows represent the source nodes, and columns represent target nodes, i.e. take the transpose.

Mathematically, we're doing

`( P4 * P3 * P2 * P1 )^T = P1^T * P2^T * P3^T * P4^T`

In implementation, this could be achieved by taking the transpose of each matrix we had been working with (e.g. P1^T ) and multiplying in the reverse order, e.g.
```
node_scores = P1^T
node_scores = node_scores @ P2^T
node_scores = node_scores @ P3^T
```
instead of
```
node_scores = P1
node_scores = P2 @ node_scores
node_scores = P3 @ node_scores
```

Rather than getting the matrix A1 then normalizing it to be P1 then taking the transpose, the efficient thing is to get the transpose of the matrix A1 from the start, i.e. re-write `metaedge_to_adjacency_matrix` to output the transpose of what it's currently outputting,
then change the way we normalize during `diffusion_step`.

If you can re-write `metaedge_to_adjacency_matrix` to output the transpose of what it's currently producing, I can alter the normalization in diffusion step.

File: hetmech/diffusion.py (hunk 1 above)
Comment (kkloste) [inline hetmech/diffusion.py]: Currently `node_scores` is initialized as a row vector, which works for our current purposes.
No need to make a change here.

**Comment by kkloste** [comment]
Uh, I inadvertently started a review. I don't know how that happened. Those were intended to be comments.

**Comment by kkloste** [comment]
This is my first time working on a PR from a fork. I'm a little confused how I should make edits to a file.
I want to make changes to `diffusion.py` -- do I access that file from here: https://github.com/dhimmel/hetmech/blob/transpose/hetmech/diffusion.py   ?
I assume I need to work on the dhimmel version of the file?

**Comment by dhimmel** [comment]
> I want to make changes to diffusion.py -- do I access that file from https://github.com/dhimmel/hetmech/blob/transpose/hetmech/diffusion.py?

@kkloste usually you would create a pull request that requests to modify my `dhimmel:transpose` branch.

But since I have "Allow edits from maintainers." selected, you should be able to click the pencil from the link you mentioned and make the changes using the GitHub UI directly to my branch. Or you can probably make a commit locally and push it to `dhimmel:transpose`.

**Comment by kkloste** [comment]
Ok, cool, thanks.

**Comment by kkloste** [comment]
We have test functions for the normalizing function, I thought?

**Comment by dhimmel** [comment]
> We have test functions for the normalizing function, I thought?

My bad.. yeah. and they're failing as they should

File: hetmech/test_diffusion.py [hunk 2]
```diff
@@ -56,7 +56,7 @@ def test_diffusion_step_row_and_column_damping(
         input_matrix = self.get_clean_matrix()
 
         # Create the matrix expected for simultaneous dual normalization
-        pr, pc = row_damping, column_damping
+        pc, pr = row_damping, column_damping
```
Comment (dhimmel) [inline hetmech/test_diffusion.py]: I made these tests pass, by this ugly hack... would it be hard to rewrite `expect` so we can get rid of the transpose below and this change?

File: hetmech/diffusion.py (hunk 1 above)
Comment (dhimmel) [inline hetmech/diffusion.py]: > No need to make a change here.

Got it, fixing `diffustion_step` was all that was necessary. That is great!

File: hetmech/test_diffusion.py (hunk 2 above)
Comment (kkloste) [inline hetmech/test_diffusion.py]: Not hard, just tedious :-)
I'll take care of it!

**Comment by dhimmel** [comment]
@kkloste thanks for fixing. I'm going to squash all the diffusion commits, but the final state will be the same

**Comment by dhimmel** [comment]
@kkloste you can merge this once you approve. Don't use squash merge... use "Create a merge commit" since I want to keep these 4 commits separate.

In the commit message body, you can add `Closes #27.`

**Comment by kkloste** [comment]
Hm. For some reason I am unable to merge. I might not have necessary permissions?
All I can do is close, so I'm closing.

**Comment by dhimmel** [comment]
> Hm. For some reason I am unable to merge. I might not have necessary permissions?

Because, it hasn't received an "Approved" review. You have to provide the review first.

**Comment by kkloste** [comment]
Ah! Thanks for the lesson :-)

**Pull request merged**